Start Surface Distance Library

MattKear opened the pull request:
Following on from the discussion in this PR: https://github.com/ArduPilot/ardupilot/pull/26827, it was agreed that the best way forward was to move the copter rangefinder functionality into a library (we called it AP_HAGL then).  This PR is the start of that process.

**No functional change should result from this PR**

As his way, @IamPete1 already had a branch for it! https://github.com/ArduPilot/ardupilot/pull/20535

I have rebased his original PR to get this.  I will continue the work and do the testing.

rmackay9 commented:
This is looking good!  Let's call the new library "A**P**_SurfaceDistance" because we will likely use it in Plane as well.

tridge commented:
maybe a STRK log message? with instance number

MattKear commented:
Key things to test are:
- Behaviour with rangefinder out of range
- Behaviour with rangefinder enable/disable on RC switch

MattKear commented:
Renamed to AP_SurfaceDistance and added logging.  For logging I have added a status bitmask which should make it easier to track various fault conditions.

timtuxworth commented:
> This is looking good! Let's call the new library "A**P**_SurfaceDistance" because we will likely use it in Plane as well.

Definitely want this in Plane!

IamPete1 commented on ArduCopter/Copter.h:
```diff
@@ -251,20 +252,10 @@ class Copter : public AP_Vehicle {
-    } rangefinder_state, rangefinder_up_state;
-
-    // return rangefinder height interpolated using inertial altitude
+    AP_SurfaceDistance rangefinder_state {ROTATION_PITCH_270, inertial_nav, 0U};
+    AP_SurfaceDistance rangefinder_up_state {ROTATION_PITCH_90, inertial_nav, 1U};
```
```suggestion
#if RANGEFINDER_ENABLED == ENABLED && AP_RANGEFINDER_ENABLED
    AP_SurfaceDistance rangefinder_state {ROTATION_PITCH_270, inertial_nav, 0U};
    AP_SurfaceDistance rangefinder_up_state {ROTATION_PITCH_90, inertial_nav, 1U};
#endif
```
if you can get it to build without too much messing about.

MattKear replied:
copter accesses the rangefinder_state too much.  I would have to do too much copter rework, don't want to change that much in this PR

IamPete1 commented on libraries/AP_SurfaceDistance/AP_SurfaceDistance.cpp:
```diff
@@ -0,0 +1,182 @@
+ # define RANGEFINDER_TIMEOUT_MS 1000        // rangefinder filter reset if no updates from sensor in 1 second
+#endif
+
+#if AP_RANGEFINDER_ENABLED
```
Should be above `RANGEFINDER_TIMEOUT_MS`

MattKear replied:
Need it outside for stale data function/check to still work.  Copter doesn't have a lot rangefinder defines in it and I am hesitant to go adding them in this pr.

IamPete1 commented on libraries/AP_SurfaceDistance/AP_SurfaceDistance.cpp:
```diff
@@ -0,0 +1,182 @@
+
+void AP_SurfaceDistance::update()
+{
+    RangeFinder *rangefinder = RangeFinder::get_singleton();
```
```suggestion
    const RangeFinder *rangefinder = RangeFinder::get_singleton();
```
If it will build.

IamPete1 commented on libraries/AP_SurfaceDistance/AP_SurfaceDistance.cpp:
```diff
@@ -0,0 +1,182 @@
+    // update health
+    alt_healthy = ((rangefinder->status_orient(rotation) == RangeFinder::Status::Good) &&
+                            (rangefinder->range_valid_count_orient(rotation) >= RANGEFINDER_HEALTH_MAX));
+    if (!alt_healthy){
```
```suggestion
    if (!alt_healthy) {
```

IamPete1 commented on libraries/AP_SurfaceDistance/AP_SurfaceDistance.cpp:
```diff
@@ -0,0 +1,182 @@
+    }
+
+    // update health
+    alt_healthy = ((rangefinder->status_orient(rotation) == RangeFinder::Status::Good) &&
+                            (rangefinder->range_valid_count_orient(rotation) >= RANGEFINDER_HEALTH_MAX));
```
```suggestion
    alt_healthy = (rangefinder->status_orient(rotation) == RangeFinder::Status::Good) &&
                            (rangefinder->range_valid_count_orient(rotation) >= RANGEFINDER_HEALTH_MAX);
```

IamPete1 commented on libraries/AP_SurfaceDistance/AP_SurfaceDistance.cpp:
```diff
@@ -0,0 +1,182 @@
+            reset_terrain_offset = true;
+            set_status(Status::Stale_Data);
+        } else {
+            alt_cm_filt.apply(alt_cm, 0.05f);
```
Not necessarily a issue here, but this is assuming a 20Hz update rate. Excitingly plane runs its rangefinders at 50Hz....

MattKear replied:
ok, did not know about that difference in rangefinder, but that can come when we work to integrate this into plane.

MattKear replied:
I have made this a more generic now, by getting last loop time from AP::scheduler.  This **should** be a negligible change to copter behaviour.

IamPete1 replied:
These do not run at loop rate, there running slower in the copter scheduler table. The 0.05 was OK, we just need to remember it when we port to another vehicle.

IamPete1 commented on libraries/AP_SurfaceDistance/AP_SurfaceDistance.cpp:
```diff
@@ -0,0 +1,182 @@
+    }
+
+    // handle reset of terrain offset
+    if (reset_terrain_offset) {
+        if (rotation == ROTATION_PITCH_90) {
+            // upward facing
+            terrain_offset_cm = inertial_alt_cm + alt_cm;
+        } else {
+            // assume downward facing
+            terrain_offset_cm = inertial_alt_cm - alt_cm;
+        }
+    }
```
I wonder if this bit is a little too vehicle specific. Maybe it would be better at the copter level still. I'm not sure. Certainly if we allow using this for surfaces that are not up or down.

MattKear replied:
sub has a similar workflow for terrain offset so I think it is fine to leave in this library

IamPete1 commented on libraries/AP_SurfaceDistance/AP_SurfaceDistance.cpp:
```diff
@@ -0,0 +1,182 @@
+        }
+    }
+
+    Log_Write();
```
```suggestion

#if HAL_LOGGING_ENABLED
    Log_Write();
#endif
```

IamPete1 commented on libraries/AP_SurfaceDistance/AP_SurfaceDistance.cpp:
```diff
@@ -0,0 +1,182 @@
+
+
+
+#endif // AP_RANGEFINDER_ENABLED
```
Should be at the end?

MattKear replied:
No, need to access the two helpers at the end of this file as copter does not define out the rangefinder state stuff currently (don't want to do that in this PR)

IamPete1 commented on libraries/AP_SurfaceDistance/AP_SurfaceDistance.h:
```diff
@@ -0,0 +1,53 @@
+    // get inertially interpolated rangefinder height
+    bool get_rangefinder_height_interpolated_cm(int32_t& ret) const;
+
+    bool enabled;                          // Not to be confused with rangefinder enabled, this state is to be set by the vehicle.
+    bool alt_healthy;                      // true if we can trust the altitude from the rangefinder
+    int16_t alt_cm;                        // tilt compensated altitude (in cm) from rangefinder
+    float inertial_alt_cm;                 // inertial alt at time of last rangefinder sample
+    uint32_t last_healthy_ms;
+    LowPassFilterFloat alt_cm_filt {0.5};  // altitude filter
+    int16_t alt_cm_glitch_protected;       // last glitch protected altitude
+    int8_t glitch_count;                   // non-zero number indicates rangefinder is glitching
+    uint32_t glitch_cleared_ms;            // system time glitch cleared
+    float terrain_offset_cm;               // filtered terrain offset (e.g. terrain's height above EKF origin)
```
It would be mice to make as many of these private as possible.

IamPete1 replied:
The eventual goal would be to hide everything, but that would need more work at the copter level.

MattKear replied:
I have now moved all that I can down to private without copter changes

IamPete1 commented on libraries/AP_SurfaceDistance/AP_SurfaceDistance.h:
```diff
@@ -0,0 +1,53 @@
+    uint32_t glitch_cleared_ms;            // system time glitch cleared
+    float terrain_offset_cm;               // filtered terrain offset (e.g. terrain's height above EKF origin)
+
+    const AP_InertialNav& inertial_nav;
+    const Rotation rotation;
```
Should be private.

IamPete1 commented on libraries/AP_SurfaceDistance/AP_SurfaceDistance.h:
```diff
@@ -0,0 +1,53 @@
+
+private:
+
+    void Log_Write(void) const;
```
Quite surprised this worked, `WriteStreaming` is not const. I guess if it builds it must be OK....

IamPete1 commented on libraries/AP_SurfaceDistance/AP_SurfaceDistance.h:
```diff
@@ -0,0 +1,53 @@
+    void Log_Write(void) const;
+
+    enum class Status {
+        Disabled = 0U,
```
Having this enumbated makes it a cross between a bitmask and a enum. I think just remove it. Once can assume that if the enable bit is not set then it must be disabled.

IamPete1 commented on libraries/AP_SurfaceDistance/AP_SurfaceDistance.h:
```diff
@@ -0,0 +1,53 @@
+    void set_status(Status state_change);
+
+    const uint8_t instance;
+    uint8_t status;
```
```suggestion
    Status status;
```

MattKear replied:
Can't do bitmasks with enum class unless I define the | operator for the class.  As I couldn't find this workflow in the rest of the code I decided to do it as I have done it.  Open to other ideas though.

IamPete1 commented on libraries/AP_SurfaceDistance/AP_SurfaceDistance.cpp:
```diff
@@ -0,0 +1,182 @@
+    // @Description: Surface distance measurement
+    // @Field: TimeUS: Time since system startup
+    // @Field: I: Instance
+    // @Field: St: Status
```
You can point the docs directly at the class 
```suggestion
    // @Field: St: Status
    // @FieldBitmaskEnum: St: Status
```
Although I suspect you will have to give it a annoyingly long name so the log parser finds the right one.

You could also build the enum for just the log rather than keep the state around all the time. EG:
https://github.com/ArduPilot/ardupilot/blob/4c95a3b6ef0db64ec21f6d6f048450cc9f03d8d1/ArduPlane/quadplane.cpp#L3615-L3639

MattKear replied:
Thanks for the example, I did not find that when I was looking.  Nice work flow, I have replicated.

clydemcqueen commented:
This is great work! I would love to see sub use this in a future PR.

MattKear commented:
I have flight tested this code (git hash: 426c9485)  All appears to working as before.

 Here is a break down of the test:
![image](https://github.com/ArduPilot/ardupilot/assets/34512430/02d05752-4ce4-48dd-80b1-b6735c65f795)

log file here: https://drive.google.com/file/d/1myPkiXPjzz0Ay9KEJwMSnXHhpa4FlaGW/view?usp=sharing

tridge commented on libraries/AP_SurfaceDistance/AP_SurfaceDistance.cpp:
```diff
@@ -0,0 +1,184 @@
+    const uint32_t now = AP_HAL::millis();
+
+    // Assemble assistance bitmask, definition is used to generate log documentation
+    enum class Surface_Distance_Status {
```
enum class Surface_Distance_Status : uint8_t {
xxxx

tridge commented on libraries/AP_SurfaceDistance/AP_SurfaceDistance.cpp:
```diff
@@ -0,0 +1,184 @@
+
+    // update health
+    alt_healthy = (rangefinder->status_orient(rotation) == RangeFinder::Status::Good) &&
+                            (rangefinder->range_valid_count_orient(rotation) >= RANGEFINDER_HEALTH_MAX);
```
RANGEFINDER_HEALTH_MIN better?

tridge commented on libraries/AP_SurfaceDistance/AP_SurfaceDistance.cpp:
```diff
@@ -0,0 +1,184 @@
+                                "TimeUS,I,St,D,FD,TO",
+                                "s#-mmm",//units multipliers format
+                                "F--BBB",
+                                "QfBhhf",
```
instance should be a B (uint8_t)

tridge replied:
lets make new log msg in meters

tridge commented on libraries/AP_SurfaceDistance/AP_SurfaceDistance.cpp:
```diff
@@ -0,0 +1,184 @@
+
+bool AP_SurfaceDistance::data_stale(void) const
+{
+    return (AP_HAL::millis() - last_healthy_ms) > RANGEFINDER_TIMEOUT_MS;
```
WITH_SEMAPHORE(sem);
return xxxx ...;

tridge commented on libraries/AP_SurfaceDistance/AP_SurfaceDistance.cpp:
```diff
@@ -0,0 +1,184 @@
+
+void AP_SurfaceDistance::update()
+{
+    const RangeFinder *rangefinder = RangeFinder::get_singleton();
```
WITH_SEMAPHORE(sem);

tridge commented on libraries/AP_SurfaceDistance/AP_SurfaceDistance.h:
```diff
@@ -0,0 +1,43 @@
+    uint32_t last_healthy_ms;
+
+    const AP_InertialNav& inertial_nav;
+    const Rotation rotation;
```
HAL_Semaphore sem;

MattKear commented:
![image](https://github.com/ArduPilot/ardupilot/assets/34512430/fe24d608-0515-4ee5-8f47-54cb71da8aaa)
Made those changes and tested logging IRL in the office.
Only had a downward facing LiDaR fitted so only the first instance of SURF is populated with data in this test.

MattKear commented:
Apologies, I cannot make the call this evening, but this PR has been through a few reviews now and I am hopeful that it is good to go.
As a reminder, since last review the following changes have been made:
-  Rename RANGEFINDER_HEALTH_MAX to RANGEFINDER_HEALTH_MIN
-  Logging has been changed to report in meters and fixed-up data types.  This has been tested.  See plot above.
- Semaphore added to avoid accessing last_healthy_ms out of sequence with millis() in data_stale() helper.

rmackay9 reviewed (approved):
Nice to see the testing results! Thanks!
